### PR TITLE
Derive instances when data types use type synonyms

### LINKED ISSUE #1443: Fail to derive Generic with record argument referenced through type synonym
``` purescript
type MyArray   = Array String
type MyString  = String
type MyNumber  = Number
type MyInt     = Int
type MyBoolean = Boolean
type MyProduct = Maybe String
type MyRecord  = { x :: Number, y :: Number }

data Foo = Foo MyArray   -- fine
               MyString  -- fine
               MyNumber  -- fine
               MyInt     -- fine
               MyBoolean -- fine
               MyProduct -- fine
               MyRecord  -- FAIL
derive instance fooGeneric :: Generic Foo
```

fails with

``` shell
No instance found for Data.Generic.Generic MyModule.MyRecord<>
```

### LINKED ISSUE #2416: Allow `Newtype` deriving for newtypes that use type synonyms

### LINKED ISSUE #2481: Allow `derive newtype instance` for newtypes that use type synonyms
My minimal reproduction is
```
type MyString = String
newtype WrappedString = WrappedString MyString
derive newtype instance semigroupWrappedString :: Semigroup WrappedString
```
which produces the error
```
    No type class instance was found for

      Data.Semigroup.Semigroup MyString


  while checking that expression #dict Semigroup
    has type { append :: WrappedString -> WrappedString -> WrappedString
             }
  in value declaration semigroupWrappedString
```
even though 
```
newtype WrappedString = WrappedString String
derive newtype instance semigroupWrappedString :: Semigroup WrappedString
```
compiles without error.

(Originally posted incorrectly to #2416, but this is a different bug.)

### PR DESCRIPTION
**Pull request by paf31**
Fixes #2481 
Fixes #2416
Fixes #1443

I would have liked to fix this by moving deriving into the type checker, but then we'd have to rearrange a whole bunch of stuff, so instead this just uses the externs to build a map of type synonym data, and then reuses the existing `replaceAllTypeSynonyms'` function to eliminate any synonyms in the instance head.

### CODE REVIEW

File: src/Language/PureScript/TypeChecker/Synonyms.hs [hunk 1]
```diff
@@ -5,43 +5,47 @@
...
+import           Language.PureScript.Environment
+import           Language.PureScript.Errors
+import           Language.PureScript.Kinds
+import           Language.PureScript.Names
+import           Language.PureScript.TypeChecker.Monad
+import           Language.PureScript.Types
+
+-- | Replace fully applied type synonyms
+replaceAllTypeSynonyms'
+  :: M.Map (Qualified (ProperName 'TypeName)) ([(Text, Maybe Kind)], Type)
```
Comment (garyb) [inline src/Language/PureScript/TypeChecker/Synonyms.hs]: Could move the `SynonymMap` type into this module, and then use it here?

File: src/Language/PureScript/Sugar/TypeClasses/Deriving.hs [hunk 2]
```diff
@@ -3,43 +3,71 @@
...
+          Just (Qualified (Just mn) name, (args, ty))
+        fromLocalDecl (PositionedDeclaration _ _ d) = fromLocalDecl d
+        fromLocalDecl _ = Nothing
+
+replaceAllTypeSynonymsM
+  :: MonadError MultipleErrors m
+  => SynonymMap
+  -> Type
+  -> m Type
+replaceAllTypeSynonymsM syns = either throwError pure . replaceAllTypeSynonyms' syns
```
Comment (garyb) [inline src/Language/PureScript/Sugar/TypeClasses/Deriving.hs]: Maybe this should go in `Language.PureScript.TypeChecker.Synonyms` too?

File: src/Language/PureScript/Sugar/TypeClasses/Deriving.hs [hunk 3]
```diff
@@ -3,43 +3,71 @@
...
 -- elaborates that into an instance declaration via code generation.
 deriveInstance
   :: (MonadError MultipleErrors m, MonadSupply m)
   => ModuleName
+  -> SynonymMap
   -> [Declaration]
   -> Declaration
   -> m Declaration
-deriveInstance mn ds (TypeInstanceDeclaration nm deps className tys@[ty] DerivedInstance)
+deriveInstance mn _ ds (TypeInstanceDeclaration nm deps className tys@[ty] DerivedInstance)
```
Comment (LiamGoodacre) [inline src/Language/PureScript/Sugar/TypeClasses/Deriving.hs]: I think these need to do synonym substitution too.  They inspect data constructor arguments; for example, `deriveOrd` calls `objectType` to check for a record.

Comment (LiamGoodacre) [inline src/Language/PureScript/Sugar/TypeClasses/Deriving.hs]: The following fail on 0.10.3:
```purescript
type L = {}
data X = X L
derive instance eqX :: Eq X

type M = {}
data Y = Y {foo :: M}
derive instance eqY :: Eq Y

type N = {}
data Z = Z N
derive instance eqZ :: Eq Z
```

Comment (paf31) [inline src/Language/PureScript/Sugar/TypeClasses/Deriving.hs]: Yes, I realized this last night after I finished working on this 😄 

I'll try to finish this up later.

**Comment by paf31** [comment]
This is ready for review again, those issues should now be fixed.

**Comment by LiamGoodacre** [comment]
Looks good to me 👍

**Comment by paf31** [comment]
Ok, I'll merge this since @LiamGoodacre has signed off and the tests are passing.

**Comment by esad** [comment]
This still doesn't work for record aliases. I can't get the @garyb example to compile with v0.10.5, it fails with `No type class instance was found for Data.Generic.Generic { "baz" :: { "foo" :: String } }`.

**Comment by paf31** [comment]
That's a different issue though. It means that we don't derive `Generic` for nested records (yet).

**Comment by esad** [comment]
@paf31 This minimal, non-nested example fails too:

```
type Baz = { baz :: String }
newtype T = T Baz

derive instance genericT :: Generic T
```

(`No type class instance was found for Data.Generic.Generic { "baz" :: String }`)

**Comment by paf31** [comment]
Then that is a bug, thanks for the report 😄

**Pull request merged**